packages/core/src/search/ir.ts

```diff
@@ -0,0 +1,244 @@
+// Search Immediate Representation (IR) module
+//
+// https://hl7.org/fhir/R4/search.html
+//
+// FHIR R4 has the following search parameter types:
+//  1. Number
+//  2. Date/DateTime
+//  3. String
+//  4. Token
+//  5. Reference
+//  6. Composite
+//  7. Quantity
+//  8. URI
+//  9. Special
+//
+// To make matters more complicated, we must consider that these search parameters can be applied
+// to many different underlying element types.
+//
+// To make our lives easier, we will use a simple Immediate Representation (IR) format to represent the search parameters.
+// All underlying element types will be mapped to the IR format for the corresponding search parameter type.
+
+import { CodeableConcept, Coding, ContactPoint, Identifier, Period, Quantity } from '@medplum/fhirtypes';
+import { isQuantity, toPeriod } from '../fhirpath/utils';
+import { typedValueToString } from '../format';
+import { isReference, PropertyType, TypedValue } from '../types';
+import { getReferenceString, isResourceWithId, isString } from '../utils';
+
+export interface SearchableToken {
+  readonly system: string | undefined;
+  readonly value: string | undefined;
+}
+
+export function convertToSearchableNumbers(typedValues: TypedValue[]): number[] {
```

Comment:
Numbers also have implicit precision similar to dates; since this is going to be a public interface, should we account for that now with e.g. `number[][]`?